### DISCUSSION TITLE
Deepsleep questions

### DISCUSSION THREAD

**Discussion by Schwenkner** [arendst/Tasmota#10605] [Q&A] [answered]
May be I have overseen something.
I'm playing a bit with deepsleep currently, it did work as expected.
But now I have a test equipment with two sensors, BME680 and CCS811. 

What I have seen is, that there are cases I sometimes get only the sensor data from BME680, not CCSS811. Sometimes I get both. What an be the reason?

And one more. I have the main menu open in a browser. Everytime it returns from deepsleep I see only the data from the BME680, never the CCS811, even if they are send via MQTT. I know it makes not much sense to have a browser connected when working with deepsleep, just for understanding.

And then in a sketch of my own I have connected VCC from the sensor to a digital output to power up the sensor only when awake from deepsleep. Can this be achieved with Tasmota also?

**Comment by barbudor** [comment, marked as answer]
Hi @Schwenkner 

So far I don't see any reason why you would miss 1 sensor except for bad connections ?

Regarding controlling a pin, you could use that as a Relay output and define use a rule `on power1#boot do power on endon` and then use `on tele-BME680 do power off endon` to turn off once the Tele message has been sent.

**Reply by Schwenkner** [reply]
Bad connection cannot be the reason, since the BME data are sent everytime, I check it with MQTT explorer.

Thanks for the rules, I will give it a try.

**Reply by barbudor** [reply]
Big mistake in what I said : While the rules works, you can't power on the BME/CS after Tasmota has started.
It must be powered during Tasmota init phased or it won't be detected.

**Comment by Schwenkner** [comment]
I have added deepsleep now to one more device. It is using a BMP280.
When I'm getting data after deepsleep, the data are incorrect, pressure shows about 700 and temperature about 27 (which is too high).
As soon as I go back from deepsleep, the values are correct. It is similar to the problem I have described above for the CCS811.
With a SHT30 I do not have seen such a problem.

And one more I have seen. When I do have a browser open  and connected to my ESP with the BMP280, as soon as I enter deepsleep command using a second browser window, the pressure and temperature values going instantly to the wrong values display everytime on awake from the deepsleep.

**Reply by barbudor** [reply]
I am using a BME280 and a TSL2561 with a DeepSleepTime of 600 and Teleperiod of 300
I've seen no problem of values. Temp & Hum values are coherent with a bluetooth Mija sensor placed side by side except for humidity slightly too high (~10%)
Pressure seems correct too but didn't tried to find an alternative source to compare.

Note that my sensors are powered normally, not through GPIO.
Are you controlling the power of your BMP ? 
It is not needed as the driver put it in low power mode before entering deepsleep

**Comment by Schwenkner** [comment]
I have had today some time for a deeper look into my deep sleep problem.
And it seems, I figured out, what is going wrong.

I have set "DeepSleepTime" from the console window, and it did work. Happy I assumed my problems are gone.
But then I've set deep sleep via MQTT, and the problem was again there.

And here are my analyse results.
Since, when I set deep sleep via MQTT, the command is published with the "retained" flag set, since I use the same command for enabling or disabling. And this is the problem. Here are the logs copied, when it awakes from deep sleep:

00:00:00 CFG: Loaded from flash at FA, Count 26
00:00:00 Project tasmota kittchen weather Version 9.2.0(sensors)-2_7_4_9(2020-12-21T15:04:56)
00:00:00 I2C: BMP280 found at 0x76
00:00:00 WIF: Connecting to AP1 Neubarnim Channel 3 BSSId DC:39:6F:47:AE:01 in mode 11n as HAS-kittchen...
00:00:02 WIF: Connected
00:00:03 HTP: Web server active on HAS-kittchen with IP address 192.168.178.47
22:12:02 MQT: Attempting connection...
22:12:02 MQT: Connected
22:12:02 MQT: HAS/Neubarnim/kittchen/tele/LWT = Online (retained)
22:12:02 MQT: HAS/Neubarnim/kittchen/cmnd/POWER = 
22:12:02 MQT: HAS/Neubarnim/kittchen/stat/RESULT = {"DeepSleepTime":120}
22:12:03 DSL: Reset wrong settings wakeup: 0, slip 10000
22:12:03 MQT: HAS/Neubarnim/kittchen/stat/RESULT = {"DeepSleep":{"Time":"2021-02-06T22:14:00","Epoch":1612646040}}

It can be seen, that it seems Tasmota gets the "retained" deep sleep command before it has sent the sensor data and goes into deep sleep. But anyway, on MQTT I've got sensor data, and they are wrong, as I said in my first post here.

As soon as I remove the "retained" message, it works.

Thus I have a workaround, but can someone try to fix this, since it would be easier to set seep sleep time then via MQTT.

**Reply by barbudor** [reply]
You shouldn't leave Deepsleeptime retained. Behavior is as coded and not a bug

The only case you would use retained is to publish DeepSleepTime 0 so the device remains awaken next time he awakes. And then clear the retained flag.

When the device receive DeepSleepTime with a value different than 0 it compute the next wake up time and goes to sleep **immediately**
As per the docs....

**Reply by Schwenkner** [reply]
Thx, after I figured out my problem, it was clear to me that it behaves like you explained.

The docs are really great, but who can read the whole (and remember each detail)?

**Comment by barbudor** [comment]
By the way, you may be interested by the brand new section on the DeepSleep page:
[Rules to adapt DeepSleep period to battery level](https://tasmota.github.io/docs/DeepSleep/#rules-to-adapt-deepsleep-period-to-battery-level)
